### PR TITLE
Add new hook to give ability to handle more mail protocols

### PR DESCRIPTION
**Pull request by cedric-anne**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #6895

Imap Oauth implementation requires a secret/token exchange between GLPI and tier providers and this part may be specific for each of them.
I propose to add a new hook that will give ability to plugins to specify a new "protocol" (e.g. "Google IMAP XOauth2") and corresponding classes to use.

Remaining work to do:
- [x] use corresponding classes;
- [x] try to see if specifying an interface is possible.

### CODE REVIEW

File: inc/auth.class.php [hunk 1]
```diff
@@ -184,14 +184,19 @@ function connection_imap($host, $login, $pass) {
             $ssl = 'TLS';
          }
 
-         $imap = new \Laminas\Mail\Protocol\Imap();
-         $imap->connect(
+         $protocol_class = Toolbox::getMailServerProtocolClassname($config['type']);
+         if ($protocol_class === null) {
+            $this->addToError(sprintf(__('Unsupported mail server type:%s.'), $config['type']));
```
Comment (trasher) [inline inc/auth.class.php:189]:
```suggestion
            thrown new \RuntimeException(sprintf(__('Unsupported mail server type:%s.'), $config['type']));
```

File: inc/toolbox.class.php [hunk 2]
```diff
@@ -1968,12 +1970,9 @@ static function parseMailServerConnectString($value, $forceport = false) {
       }
       $tab['mailbox'] = preg_replace("/.*}/", "", $value);
 
-      $tab['type']    = '';
-      if (strstr($value, "/imap")) {
-         $tab['type'] = 'imap';
-      } else if (strstr($value, "/pop")) {
-         $tab['type'] = 'pop';
-      }
+      $type = preg_replace('/^\{[^\/]+\/([^\/]+)(?:\/.+)*\}.*/', '$1', $value);
```
Comment (trasher) [inline inc/toolbox.class.php:1973]: I do not speak regexp fluently, please add a comment

**Comment by trasher** [comment]
I guess it should be possible to unit test at least the new methods (getMailServerProtocols etc); espacially to be sure core protocls cannot be overriden, and new ones are really loaded

File: inc/toolbox.class.php [hunk 3]
```diff
@@ -1968,12 +1970,12 @@ static function parseMailServerConnectString($value, $forceport = false) {
       }
       $tab['mailbox'] = preg_replace("/.*}/", "", $value);
 
-      $tab['type']    = '';
-      if (strstr($value, "/imap")) {
-         $tab['type'] = 'imap';
-      } else if (strstr($value, "/pop")) {
-         $tab['type'] = 'pop';
-      }
+      // type follows first found "/" adn ends on next "/" (or end of server string)
```
Comment (trasher) [inline inc/toolbox.class.php:1973]:
```suggestion
      // type follows first found "/" and ends on next "/" (or end of server string)
```

**Comment by carlosbmc** [comment]
When will be this new feature available?

**Pull request merged**